### PR TITLE
Better dataset and data documentation

### PR DESCRIPTION
**Pull request by joecummings**
#### Context
- This addresses some concerns about documentation for our datasets.

#### Changelog
- Expose dataset and data classes
- Update docstrings

#### Test plan
- Eyes

### CODE REVIEW

File: torchtune/datasets/_chat.py [hunk 1]
```diff
@@ -28,11 +28,12 @@ class ChatDataset(Dataset):
...
+    If the column/key names differ from the expected names in the ``ChatFormat``,
+    then the ``column_map`` argument can be used to provide this mapping.
+
+    Use ``convert_to_messages`` to prepare your dataset into the llama conversation format
+    and roles::
 
-    Use `convert_to_messages` to prepare your dataset into the llama conversation format
-    and roles:
         [
             {
```
Comment (RdoubleA) [inline torchtune/datasets/_chat.py:38]: while you're here, can you also update this format example? It should actually be something like:
```
[
    Message(
        role=<system|user|assistant>,
        content=<message>,
    ),
    ...
]
```

File: docs/source/api_ref_data.rst [hunk 2]
```diff
@@ -14,8 +17,36 @@ torchtune.data
...
     MistralChatFormat
+
+Types
+-----
+
+.. autosummary::
+    :toctree: generated/
+    :nosignatures:
+
+    Message
```
Comment (RdoubleA) [inline docs/source/api_ref_data.rst:42]: I think sphinx breaks for Message because it does not have docstrings

**Comment by RdoubleA** [comment]
I'm betting you'll have to fix a lot of sphinx errors in the docstrings, so I will stamp once that passes

File: docs/source/api_ref_data.rst (hunk 2 above)
Comment (joecummings) [reply docs/source/api_ref_data.rst:42]: Message is fine, it broke for ``truncate``

File: torchtune/datasets/_chat.py (hunk 1 above)
Comment (joecummings) [reply torchtune/datasets/_chat.py:38]: Wait so the user has to create a `Message` class? They can't just have a JSONL file that parses this into a Message format?

Comment (RdoubleA) [reply torchtune/datasets/_chat.py:38]: hm yeah, that would be more ideal... maybe we should add this step in the dataset class to parse it into Message format for you if needed

Comment (joecummings) [reply torchtune/datasets/_chat.py:38]: Well we *could* enforce it here on the class, but in the builder, it will format for you?

Comment (RdoubleA) [reply torchtune/datasets/_chat.py:38]: yeah if you specify a converter... need to think more on this, let's revisit this.

**Review by RdoubleA — APPROVED** [review]

File: torchtune/datasets/_chat.py (hunk 1 above)
Comment (joecummings) [reply torchtune/datasets/_chat.py:38]: Definitely, but for now, what should the docstring read?

**Pull request merged**